cpp/lib/Authenticator.cpp

```diff
@@ -356,12 +356,197 @@ namespace Client
+      m_disableSamlUrlCheck = m_connection->disable_saml_url_check;
+      m_retriedCount = get_login_retry_count(m_connection);
+      m_retryTimeout = get_retry_timeout(m_connection);
+
+      //m_appID = m_connection->application_name;
+      //m_appVersion = m_connection->application_version;
+      m_appID = "ODBC";
+      m_appVersion = "3.4.1";
+  }
+
+  AuthenticatorOKTA::~AuthenticatorOKTA()
+  {
+      // nop
+  }
+
+  void AuthenticatorOKTA::curl_get_call(SFURL& url, jsonObject_t& resp, bool parseJSON, std::string& rawData)
+  {
+      bool isRetry = false;
+      int64 maxRetryCount = get_login_retry_count(m_connection);
+      int64 elapsedTime = 0;
+      int64 renewTimeout = auth_get_renew_timeout(m_connection);
+      int64 curlTimeout = m_connection->network_timeout;
+
+      std::string destination = url.toString();
+      void* curl_desc;
+      CURL* curl;
+      curl_desc = get_curl_desc_from_pool(destination.c_str(), m_connection->proxy, m_connection->no_proxy);
+      curl = get_curl_from_desc(curl_desc);
+
+      SF_ERROR_STRUCT *err = &m_connection->error;
+
+      NON_JSON_RESP* raw_resp = new NON_JSON_RESP;
+      raw_resp->write_callback = non_json_resp_write_callback;
+      RAW_JSON_BUFFER buf = { NULL,0 };
+      raw_resp->buffer = (void*)&buf;
+
+      // add headers for account and authentication
+      SF_HEADER* httpExtraHeaders = sf_header_create();
+
+      httpExtraHeaders->use_application_json_accept_type = SF_BOOLEAN_TRUE;
+      if (!create_header(m_connection, httpExtraHeaders, &m_connection->error)) {
+          log_trace("sf", "AuthenticatorOKTA",
+              "get_curl_call",
+              "Failed to create the header for the request to get onetime token");
+          SET_SNOWFLAKE_ERROR(err, SF_STATUS_ERROR_GENERAL, "OktaConnectionFailed: failed to create the header", SF_SQLSTATE_GENERAL_ERROR);
+          goto cleanup;
+      }
+
+      if (!http_perform(curl, GET_REQUEST_TYPE, (char*)destination.c_str(), httpExtraHeaders, NULL, NULL, raw_resp,
+              curlTimeout, SF_BOOLEAN_FALSE, err,
+              m_connection->insecure_mode, m_connection->ocsp_fail_open,
+              m_connection->retry_on_curle_couldnt_connect_count,
+              renewTimeout, maxRetryCount, &elapsedTime, &m_retriedCount, NULL, SF_BOOLEAN_FALSE,
+              m_connection->proxy, m_connection->no_proxy, SF_BOOLEAN_FALSE, SF_BOOLEAN_FALSE)) 
+      {
+          //Fail to get the saml response. Retry.
+              isRetry = true;
+              goto cleanup;
+      }
+
+      if (elapsedTime >= m_retryTimeout)
+      {
+          CXX_LOG_WARN("sf", "AuthenticatorOKTA", "get_curl_call",
+              "Fail to get SAML response, timeout reached: %d, elapsed time: %d",
+              m_retryTimeout, elapsedTime);
+
+          SET_SNOWFLAKE_ERROR(err, SF_STATUS_ERROR_REQUEST_TIMEOUT, "OktaConnectionFailed: timeout reached", SF_SQLSTATE_GENERAL_ERROR);
+          goto cleanup;
```

Comment:
Please don't use goto in C++ code. Use RAII and if you need to manage resource from C API use unique_ptr (https://en.cppreference.com/w/cpp/memory/unique_ptr/unique_ptr constructors (3) and (4))

Reply:
I use the exception instead of goto. In my opinion, using unique_ptr is inappropriate for this because I should move the pointer to call the function and get the pointer back from the HTTP perform.